docker_kernel/kernel.py

```diff
@@ -47,6 +47,11 @@ def __init__(self, *args, **kwargs):
         self._build_stage_aliases = {}
         self._build_context_dir: str = os.getcwd() # directory the kernel was started in
         self._frontend = None
+        self._tmp_dir = tempfile.TemporaryDirectory()
+        copy_files(self._build_context_dir, self._tmp_dir.name)
+
+    def __del__(self):
```

Comment:
Is the `__del__ method` necessary, as I find `remove_tmp_dir` actually more explanatory?

Reply:
`__del__` is a finalizer method. It gets called when the object is garbage collected. So this basically ensures that `remove_temp_dir` gets called when the Python interpreter destroys our Kernel

Reply:
Okay I did some more research on this and found [this](https://stackoverflow.com/questions/865115/how-do-i-correctly-clean-up-a-python-object). So it seems like we have two options:
1. Use `atexit.register` to execute code on program termination
2. Use `__enter__` and `__exit__` as well as `with ... as ...` where our kernel is created to ensure deletion of the kernel triggers our cleanup.

But I guess we basically need to do option 2 because the user could potentially instantiate multiple kernels by opening multiple notebooks. Option 1 can only clean up ONCE when the whole program terminates.